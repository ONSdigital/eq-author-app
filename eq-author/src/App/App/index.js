import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { injectGlobal, ThemeProvider } from "styled-components";
import theme, { colors } from "constants/theme";

import { lostConnection, gainConnection } from "redux/saving/actions";

/* eslint no-unused-expressions: 0 */
injectGlobal`
  html {
    box-sizing: border-box;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  html,
  body {
    height: 100%;
    width: 100%;
    font-size: 1em;
    margin: 0;
    padding: 0;
    -webkit-font-smoothing: antialiased;
  }

  body {
    font-family: 'Lato', sans-serif;
    overflow: hidden;
    color: ${colors.text};
  }

  input,
  select,
  textarea,
  button {
    font-family: inherit;
    -webkit-font-smoothing: inherit;
  }

  a {
    color: ${colors.blue};

    &:hover {
      text-decoration: none;
    }
  }
`;

export class UnconnectedApp extends React.Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    gainConnection: PropTypes.func.isRequired,
    lostConnection: PropTypes.func.isRequired
  };

  componentDidMount = () => {
    window.addEventListener("online", this.props.gainConnection);
    window.addEventListener("offline", this.props.lostConnection);
  };

  componentWillUnmount = () => {
    window.removeEventListener("online", this.props.gainConnection);
    window.removeEventListener("offline", this.props.lostConnection);
  };

  render() {
    return <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>;
  }
}

export default connect(
  null,
  {
    lostConnection,
    gainConnection
  }
)(UnconnectedApp);
