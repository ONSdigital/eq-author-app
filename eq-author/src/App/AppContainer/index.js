/* eslint-disable react/jsx-no-bind */

import React from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import { AppContainer } from "react-hot-loader";

import { Switch } from "react-router-dom";
import { Route, Router } from "react-router";
import { ApolloProvider } from "react-apollo";
import PrivateRoute from "App/components/PrivateRoute";
import RedirectRoute from "App/components/RedirectRoute";
import { Routes as RoutePaths } from "utils/UrlUtils";

import QuestionnairesPage from "App/QuestionnairesPage";
import SignInPage from "App/SignInPage";
import QuestionnaireDesignPage from "App/QuestionPage/Design/QuestionnaireDesignPage";
import NotFoundPage from "App/NotFoundPage";
import { Provider, connect } from "react-redux";
import { isSignedIn } from "redux/auth/reducer";

export const Routes = ({ isSignedIn, ...otherProps }) => (
  <Router {...otherProps}>
    <Switch>
      <Route path={RoutePaths.SIGN_IN} component={SignInPage} exact />
      <PrivateRoute
        path={RoutePaths.HOME}
        component={QuestionnairesPage}
        isSignedIn={isSignedIn}
        exact
      />
      <RedirectRoute
        from="/questionnaire/:questionnaireId/design/:sectionId/:pageId"
        to={RoutePaths.PAGE}
      />
      <RedirectRoute
        from="/questionnaire/:questionnaireId/design/:sectionId"
        to={RoutePaths.SECTION}
      />
      <PrivateRoute
        path={RoutePaths.QUESTIONNAIRE}
        exact={false}
        component={QuestionnaireDesignPage}
        isSignedIn={isSignedIn}
      />
      <Route path="*" component={NotFoundPage} exact />
    </Switch>
  </Router>
);

Routes.propTypes = {
  isSignedIn: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
  isSignedIn: isSignedIn(state)
});
const ConnectedRoutes = connect(mapStateToProps)(Routes);

const App = ({ store, client, history }) => {
  return (
    <AppContainer>
      <ApolloProvider client={client}>
        <Provider store={store}>
          <ConnectedRoutes history={history} />
        </Provider>
      </ApolloProvider>
    </AppContainer>
  );
};

App.propTypes = {
  client: CustomPropTypes.apolloClient.isRequired,
  store: CustomPropTypes.store.isRequired,
  history: CustomPropTypes.history.isRequired
};

export default App;
