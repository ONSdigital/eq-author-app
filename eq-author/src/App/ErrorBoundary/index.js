import React from "react";
import PropTypes from "prop-types";
import NotFound from "App/NotFoundPage";
import { ERR_PAGE_NOT_FOUND } from "constants/error-codes";

export default class ErrorBoundary extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired
  };

  state = { error: false };

  componentDidCatch(error) {
    this.setState({ error });
  }

  render() {
    if (this.state.error.message === ERR_PAGE_NOT_FOUND) {
      // You can render any custom fallback UI
      return <NotFound />;
    }

    return this.props.children;
  }
}
