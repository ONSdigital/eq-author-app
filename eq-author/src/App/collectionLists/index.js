import React from "react";
import { Route } from "react-router-dom";
import CollectionListsPage from "./collectionListsPage";

export default [
  <Route
    key="collectionLists"
    path="/q/:questionnaireId/collectionLists"
    render={() => <CollectionListsPage />}
  />,
];
