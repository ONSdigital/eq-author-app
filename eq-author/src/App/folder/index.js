import React from "react";

import { Route, Redirect } from "react-router-dom";
import SkipLogic from "App/shared/Logic/SkipLogic";

import Design from "./Design";
import RoutingPage from "./Routing";

export default [
  <Route
    key="folder-design"
    path="/q/:questionnaireId/folder/:folderId/design"
    component={Design}
  />,
  <Route key="folder-logic" path="/q/:questionnaireId/folder/:folderId/logic">
    <Redirect to="routing" />
  </Route>,
  <Route
    key="folder-logic-routing"
    path="/q/:questionnaireId/folder/:folderId/routing"
    component={RoutingPage}
  />,
  <Route
    key="folder-logic-skip"
    path="/q/:questionnaireId/folder/:folderId/skip"
    component={SkipLogic}
  />,
];
