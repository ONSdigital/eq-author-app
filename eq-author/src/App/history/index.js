import React from "react";
import { Route } from "react-router-dom";
import HistoryPage from "./HistoryPage";

export default [
  <Route
    key="history"
    path="/q/:questionnaireId/history"
    component={HistoryPage}
  />,
];
