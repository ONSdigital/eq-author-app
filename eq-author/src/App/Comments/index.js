import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import { useQuery, useMutation } from "@apollo/react-hooks";
import { useSubscription } from "react-apollo";
import { useMe } from "App/MeContext";

import COMMENT_QUERY from "./graphql/commentsQuery.graphql";

import COMMENT_ADD from "./graphql/createNewComment.graphql";
import REPLY_ADD from "./graphql/createNewReply.graphql";

import COMMENT_SUBSCRIPTION from "./graphql/commentSubscription.graphql";

import Error from "components/Error";
import Loading from "components/Loading";
import Comment from "components/Comments/Comment";
import CommentEditor from "components/Comments/CommentEditor";
import Collapsible from "components/Collapsible";

const Wrapper = styled.section`
  h1 {
    font-size: 1em;
    margin: 0;
    padding: 1em;
    padding-bottom: 0;
  }

  > ul > li,
  .AddComment {
    padding: 1em;
    padding-bottom: 0;
  }

  > ul {
    padding: 0;

    > li {
      display: block;
      border-top: 1px solid rgb(228, 232, 235);
    }
  }
`;

const Replies = styled(Collapsible)`
  .collapsible-title,
  .collapsible-title > * {
    font-weight: normal;
    padding: 0;
  }

  .collapsible-header {
    margin-bottom: 1em;
  }

  .collapsible-body {
    border-left: none;
    margin-top: 0;
  }

  ul {
    padding: 0;

    li {
      display: block;
    }
  }
`;

const CommentsPanel = ({ componentId }) => {
  const [createComment] = useMutation(COMMENT_ADD);
  const [createReply] = useMutation(REPLY_ADD);

  const { loading, error, data, refetch } = useQuery(COMMENT_QUERY, {
    variables: {
      componentId,
    },
  });

  useSubscription(COMMENT_SUBSCRIPTION, {
    variables: {
      id: componentId,
    },
    onSubscriptionData: () => {
      refetch();
    },
  });

  const { me } = useMe();

  if (loading) {
    return <Loading height="100%">Comments loadingâ€¦</Loading>;
  }

  if (error) {
    return <Error>Oops! Something went wrong</Error>;
  }

  const { comments } = data;

  const formatName = (name) =>
    name.replace(/\w\S*/g, (w) => w.replace(/^\w/, (c) => c.toUpperCase()));

  const sortByDate = (commentArr) =>
    commentArr.sort((a, b) =>
      b.datePosted.toString().localeCompare(a.datePosted.toString())
    );

  const formatComment = ({
    id,
    user,
    commentText,
    createdTime,
    editedTime,
  }) => ({
    id,
    authorName: formatName(user.displayName) || user.name || "",
    canEdit: user.id === me.id,
    canDelete: user.id === me.id,
    datePosted: createdTime,
    dateModified: editedTime,
    commentText,
  });

  const formattedComments = sortByDate(
    comments.map(({ replies, ...rest }) => ({
      ...formatComment(rest),
      replies: sortByDate(
        replies.map((comment) => ({
          ...formatComment(comment),
        }))
      ),
    }))
  );

  const renderReplies = (replies, rootId, componentId) => (
    <ul>
      {replies.map(({ id: replyId, ...rest }) => (
        <li key={`comment-reply-${replyId}`}>
          <Comment
            {...rest}
            id={replyId}
            rootId={rootId}
            subjectId={componentId}
            isReply
          />
        </li>
      ))}
    </ul>
  );

  const renderComments = (comments = [], componentId) => (
    <ul>
      {comments.map(({ id: commentId, replies, ...rest }) => (
        <li key={`comment-${commentId}`} data-test={`Comment-${commentId}`}>
          <Comment
            {...rest}
            id={commentId}
            rootId={commentId}
            subjectId={componentId}
          />
          <CommentEditor
            canClose
            startClosed
            onConfirm={(commentText) =>
              createReply({
                variables: {
                  input: {
                    componentId,
                    commentId,
                    commentText,
                  },
                },
              })
            }
          />
          {replies.length > 0 && (
            <Replies
              showHide
              title={`${replies.length} ${
                replies.length > 1 ? "replies" : "reply"
              }`}
              withoutHideThis
            >
              {renderReplies(replies, commentId, componentId)}
            </Replies>
          )}
        </li>
      ))}
    </ul>
  );

  return (
    <Wrapper data-test="comments-panel">
      <h1>Comments</h1>
      <CommentEditor
        key={`add-comment-${componentId}`}
        className="AddComment"
        onConfirm={(commentText) =>
          createComment({
            variables: {
              input: {
                componentId,
                commentText,
              },
            },
          })
        }
      />
      {renderComments(formattedComments, componentId)}
    </Wrapper>
  );
};

CommentsPanel.propTypes = {
  /**
   * The ID of the component users are commenting on. This may be a page, question page, calculated summary, etc.
   */
  componentId: PropTypes.string.isRequired,
};

export default CommentsPanel;
