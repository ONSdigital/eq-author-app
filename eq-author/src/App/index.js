/* eslint-disable react/jsx-no-bind */

import React from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import { AppContainer } from "react-hot-loader";
import { Switch } from "react-router-dom";
import { Route, Router } from "react-router";
import { ApolloProvider } from "react-apollo";
import { Provider, connect } from "react-redux";

import { isSignedIn } from "redux/auth/reducer";
import PrivateRoute from "components/PrivateRoute";
import RedirectRoute from "components/RedirectRoute";
import { Routes as RoutePaths } from "utils/UrlUtils";

import QuestionnairesPage from "./QuestionnairesPage";
import SignInPage from "./SignInPage";
import QuestionnaireDesignPage from "./QuestionnaireDesignPage";
import NotFoundPage from "./NotFoundPage";
import ErrorBoundary from "./ErrorBoundary";

export const Routes = ({ isSignedIn, ...otherProps }) => (
  <Router {...otherProps}>
    <ErrorBoundary>
      <Switch>
        <Route path={RoutePaths.SIGN_IN} component={SignInPage} exact />
        <PrivateRoute
          path={RoutePaths.HOME}
          component={QuestionnairesPage}
          isSignedIn={isSignedIn}
          exact
        />
        <RedirectRoute
          from="/questionnaire/:questionnaireId/design/:sectionId/:pageId"
          to={"/q/:questionnaireId/page/:pageId/design"}
        />
        <RedirectRoute
          from="/questionnaire/:questionnaireId/design/:sectionId"
          to={"/q/:questionnaireId/section/:sectionId/design"}
        />
        <PrivateRoute
          path={RoutePaths.QUESTIONNAIRE}
          exact={false}
          component={QuestionnaireDesignPage}
          isSignedIn={isSignedIn}
        />
        <Route path="*" component={NotFoundPage} exact />
      </Switch>
    </ErrorBoundary>
  </Router>
);

Routes.propTypes = {
  isSignedIn: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  isSignedIn: isSignedIn(state),
});
const ConnectedRoutes = connect(mapStateToProps)(Routes);

const App = ({ store, client, history }) => {
  return (
    <AppContainer>
      <ApolloProvider client={client}>
        <Provider store={store}>
          <ConnectedRoutes history={history} />
        </Provider>
      </ApolloProvider>
    </AppContainer>
  );
};

App.propTypes = {
  client: CustomPropTypes.apolloClient.isRequired,
  store: CustomPropTypes.store.isRequired,
  history: CustomPropTypes.history.isRequired,
};

export default App;
