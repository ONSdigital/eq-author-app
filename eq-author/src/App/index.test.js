import React from "react";

import App, { Routes } from "./";
import { shallow } from "enzyme";

import { createHashHistory } from "history";

const history = createHashHistory();

const client = {
  query: jest.fn(),
  readQuery: jest.fn(),
};

describe("containers/AppContainer", () => {
  it("should render", () => {
    const wrapper = shallow(<App history={history} client={client} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe("Routes", () => {
    it("should render ", () => {
      const wrapper = shallow(<Routes isSignedIn history={history} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
