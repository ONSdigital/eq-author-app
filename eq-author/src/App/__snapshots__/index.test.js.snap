// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/AppContainer Routes should render  1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  isSignedIn={true}
>
  <withApollo(withRouter(ContextProvider))>
    <ErrorBoundary>
      <Toasts>
        <Switch>
          <Route
            component={[Function]}
            exact={true}
            path="/sign-in"
          />
          <InnerComponent
            component={[Function]}
            exact={true}
            path="/"
          />
          <RedirectRoute
            from="/questionnaire/:questionnaireId/design/:sectionId/:pageId"
            to="/q/:questionnaireId/page/:pageId/design"
          />
          <RedirectRoute
            from="/questionnaire/:questionnaireId/design/:sectionId"
            to="/q/:questionnaireId/section/:sectionId/design"
          />
          <InnerComponent
            component={[Function]}
            exact={false}
            path="/q/:questionnaireId/:entityName?/:entityId?/:tab?/:modifier?"
          />
          <Route
            component={[Function]}
            exact={true}
            path="*"
          />
        </Switch>
      </Toasts>
    </ErrorBoundary>
  </withApollo(withRouter(ContextProvider))>
</Router>
`;

exports[`containers/AppContainer should render 1`] = `
<AppContainer>
  <ApolloProvider
    client={
      Object {
        "query": [MockFunction],
        "readQuery": [MockFunction],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [MockFunction],
          "getState": [MockFunction],
          "subscribe": [MockFunction],
        }
      }
    >
      <Routes
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      />
    </Provider>
  </ApolloProvider>
</AppContainer>
`;
